File: services/graphql-tms/src/mutation/geofences.rs
use async_graphql::{Context, InputObject, Object};
use uuid::Uuid;

use crate::models::geofences;

#[derive(Debug, Clone, InputObject)]
pub struct CreateGeofenceInput {
    pub name: String,
    pub coordinates: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct Mutation;

#[Object(name = "TmsGeofencesMutation")]
impl Mutation {
    async fn create_geofence(
        &self,
        ctx: &Context<'_>,
        payload: CreateGeofenceInput,
    ) -> async_graphql::Result<geofences::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, geofences::Model>(
            "insert into tms.geofences (name, coordinates) values ($1, $2) returning *",
        )
        .bind(payload.name)
        .bind(payload.coordinates)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_geofence(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        payload: CreateGeofenceInput,
    ) -> async_graphql::Result<geofences::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        let result = sqlx::query_as::<_, geofences::Model>(
            "update tms.geofences set name = $1, coordinates = $2 where id = $3 returning *",
        )
        .bind(payload.name)
        .bind(payload.coordinates)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn remove_geofence(&self, ctx: &Context<'_>, id: Uuid) -> async_graphql::Result<String> {
        let db = ctx.data::<sqlx::PgPool>()?;
        let result = sqlx::query("delete from tms.geofences where id = $1")
            .bind(id)
            .execute(db)
            .await?;

        if result.rows_affected() == 0 {
            return Err("Unable to remove geofence".into());
        }

        Ok("Geofence removed".into())
    }
}

File: src/graphql/mutations/tms/geofences.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// GEOFENCE MUTATIONS
// ============================================================================

/**
 * Creates a new geofence.
 * @param payload The input data for creating the geofence.
 */
export const createGeofence = graphql(`
  mutation CreateGeofence($payload: CreateGeofenceInput!) {
    tms {
      createGeofence(payload: $payload) {
        id
        name
        coordinates
        createdAt
        updatedAt
      }
    }
  }
`);

/**
 * Updates an existing geofence.
 * @param id The UUID of the geofence to update.
 * @param payload The input data for updating the geofence.
 */
export const updateGeofence = graphql(`
  mutation UpdateGeofence($id: UUID!, $payload: CreateGeofenceInput!) {
    tms {
      updateGeofence(id: $id, payload: $payload) {
        id
        name
        coordinates
        createdAt
        updatedAt
      }
    }
  }
`);

/**
 * Removes a geofence by its ID.
 * @param id The UUID of the geofence to remove.
 */
export const removeGeofence = graphql(`
  mutation RemoveGeofence($id: UUID!) {
    tms {
      removeGeofence(id: $id)
    }
  }
`);

/**
 * Creates a new geofence event.
 * @param payload The input data for creating the geofence event.
 */
export const createGeofenceEvent = graphql(`
  mutation CreateGeofenceEvent($payload: CreateGeofenceEventInput!) {
    tms {
      createGeofenceEvent(payload: $payload) {
        id
        eventType
        timestamp
      }
    }
  }
`);

/**
 * Removes a geofence event by its ID.
 * @param id The UUID of the geofence event to remove.
 */
export const removeGeofenceEvent = graphql(`
  mutation RemoveGeofenceEvent($id: UUID!) {
    tms {
      removeGeofenceEvent(id: $id)
    }
  }
`);
